src/redprl/refiner_types.fun

```diff
@@ -807,35 +813,35 @@ struct
         (* these operations could be expensive *)
         val Syn.TUPLE map0 = Syn.out tuple0
         val Syn.TUPLE map1 = Syn.out tuple1
-        val Syn.RECORD map = Syn.out record
-        val map0 = LabelDict.toList map0
-        val map1 = LabelDict.toList map1
-        val map = LabelDict.toList map
-
-        fun goLabel (((lbl0, a0), (lbl1, a1)), (lbl, ty)) =
-          let
-            val () = Assert.labelEq "Record.Eq" (lbl0, lbl1)
-            val () = Assert.labelEq "Record.Eq" (lbl0, lbl)
-          in
-            makeEq (I, H) ((a0, a1), ty)
-          end
-        val rec goLabels =
-          fn (([], []), []) => []
-          | ((lbla0 :: map0, lbla1 :: map1), lbla :: map)
-            => goLabel ((lbla0, lbla1), lbla) :: goLabels ((map0, map1), map)
-          | _ => raise E.error [Fpp.text "Expected the same number of labels"]
-        val goals = goLabels ((map0, map1), map)
+        val Syn.RECORD fields = Syn.out record
+
+        val {goals, famGoals, ...} = 
+          List.foldl
+            (fn (((lbl, var), ty), {goals, famGoals, env, hyps, isFirst}) =>
+               let
+                 val ty' = substVarenv env ty
+                 val m0 = Syn.LabelDict.lookup map0 lbl
+                 val m1 = Syn.LabelDict.lookup map1 lbl
+                 val env' = Var.Ctx.insert env var m0
+                 val goals' = goals >: makeEq (I, H) ((m0, m1), ty')
+                 val hyps' = hyps @> (var, CJ.TRUE ty)
+                 val famGoals' = if isFirst then famGoals else famGoals >: makeType (I, hyps) ty
+               in
+                 {goals = goals', famGoals = famGoals', env = env', hyps = hyps', isFirst = false}
+               end)
+            {goals = Lcf.Tl.empty, famGoals = Lcf.Tl.empty, env = Var.Ctx.empty, hyps = H, isFirst = true}
+            fields
```

Comment:
@favonia If I understand your question correctly, the answer is that I am moving toward an extensional treatment of records, under which we require that the specified fields exist, but do not penalize the existence of random other garbage in the record.

Record values under the new regime are un-ordered, and should be viewed in an extensional sense as finite maps that only need to obey the specification given by the record telescope.